shared/src/simulation/behaviors/request.spec.ts

```diff
@@ -0,0 +1,853 @@
+                        });
+
+                        it('should not change its state', () => {
+                            const { beforeBehaviorState, afterBehaviorState } =
+                                setupStateAndInteract(
+                                    setBehaviorState[state],
+                                    addRequestsAndPromises[requestsAndPromises],
+                                    sendEvent.resourceRequiredEvent
+                                );
+
+                            assertSameState(
+                                beforeBehaviorState,
+                                afterBehaviorState
+                            );
+                        });
+                    });
+
+                    describe('on a resource required event with a known key', () => {
+                        it('should overwrite any existing requests', () => {
+                            const { afterBehaviorState } =
+                                setupStateAndInteract(
+                                    setBehaviorState[state],
+                                    addRequestsAndPromises[requestsAndPromises],
+                                    sendEvent.resourceRequiredEventWithKnownKey
+                                );
+
+                            expect(
+                                Object.keys(
+                                    afterBehaviorState.requestedResources
+                                ).length
+                            ).toBe(1);
+                        });
+                    });
+
+                    describe.each(
+                        StrictObject.keys(sendEvent).filter((type) =>
+                            vehicleSendEvents.has(type)
+                        )
+                    )('on a %s', (event) => {
+                        it('should note the promise', () => {
+                            const { afterBehaviorState } =
+                                setupStateAndInteract(
+                                    setBehaviorState[state],
+                                    addRequestsAndPromises[requestsAndPromises],
+                                    sendEvent[event]
+                                );
+
+                            const promisedResources =
+                                afterBehaviorState.promisedResources;
+                            expect(
+                                promisedResources.length
+                            ).toBeGreaterThanOrEqual(1);
+                            const promise = promisedResources.at(-1)!;
+                            expect(promise.resource).toEqual(
+                                VehicleResource.create({ KTW: 1 })
+                            );
+                        });
+
+                        if (
+                            state === 'waiting' &&
+                            event === 'vehiclesSendEventForAnswerKey'
+                        ) {
```

Comment:
I find it extremely hard to understand the tests with the large amount of conditionals in the test logic.

As you've already created functions that initialize the state I think it should be possible to have less conditionals without causing to much duplicated code.